### PR TITLE
Add non-inline UMR

### PR DESCRIPTION
**Pull request by lappazos**

### CODE REVIEW

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 1]
```diff
@@ -6,6 +6,312 @@
...
+{
+    __be32 first_dword  = htobe32(be64toh(val) >> 32);
+    __be32 second_dword = htobe32(be64toh(val));
+
+    /* The WC spinlock, by definition, provides global ordering for all UC
+	   and WC stores within the critical region. */
+    mmio_wc_spinlock(&mmio_spinlock);
+
+    mmio_write32_be(addr, first_dword);
+    mmio_write32_be(addr + 4, second_dword);
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: cast addr to ptrdiff_t when doing address arithmetics: it is not defined in C for void* - not portable.

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 2]
```diff
@@ -6,6 +6,312 @@
...
+
+static __attribute__((constructor)) void lock_constructor(void)
+{
+    pthread_spin_init(&mmio_spinlock, PTHREAD_PROCESS_PRIVATE);
+}
+
+static inline void mmio_wc_spinlock(pthread_spinlock_t *lock)
+{
+    pthread_spin_lock(lock);
+    //#if !defined(__i386__) && !defined(__x86_64__)
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c:22]: if commented code not needed - remove or leave meaning full comment w/o #ifdefs

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 3]
```diff
@@ -6,6 +6,312 @@
...
+
+static void pthread_mmio_write64_be(void *addr, __be64 val)
+{
+    __be32 first_dword  = htobe32(be64toh(val) >> 32);
+    __be32 second_dword = htobe32(be64toh(val));
+
+    /* The WC spinlock, by definition, provides global ordering for all UC
+	   and WC stores within the critical region. */
+    mmio_wc_spinlock(&mmio_spinlock);
+
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: extra space

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 4]
```diff
@@ -6,6 +6,312 @@
...
+    mmio_write32_be(addr, first_dword);
+    mmio_write32_be(addr + 4, second_dword);
+
+    mmio_wc_spinunlock(&mmio_spinlock);
+}
+
+#define HAVE_FUNC_ATTRIBUTE_IFUNC 1
+
+#if HAVE_FUNC_ATTRIBUTE_IFUNC
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c:53]: why need this #if ? you have defined the variable 1 line above - it is always true

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 5]
```diff
@@ -6,6 +6,312 @@
...
+        __attribute__((ifunc("resolve_mmio_write64_be")));
+static write64_fn_t resolve_mmio_write64_be(void);
+#else
+__asm__(".type mmio_write64_be, %gnu_indirect_function");
+write64_fn_t resolve_mmio_write64_be(void) __asm__("mmio_write64_be");
+#endif
+
+write64_fn_t resolve_mmio_write64_be(void)
+{
+    return &pthread_mmio_write64_be;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c:64]: looks like you always resolve to your custom pthread_mmio_write64_be? Why to have all this indirection and not use that function directly for better readibility?

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 6]
```diff
@@ -6,6 +6,312 @@
...
+}
+
+xccl_status_t
+xccl_mhba_ibv_qp_to_mlx5dv_qp(struct ibv_qp *umr_qp, struct internal_qp *mqp)
+{
+    struct mlx5dv_obj dv_obj = {};
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: plz remove empty initializer ={} - it is not C standard (will give your warning with -pedantic). you memset the struct to 0 next line anyways.

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 7]
```diff
@@ -6,6 +6,312 @@
...
+    //	/* For x86 the serialization within the spin lock is enough to
+    //	 * strongly order WC and other memory types. */
+    //	mmio_wc_start();
+    //#endif
+}
+
+static inline void mmio_wc_spinunlock(pthread_spinlock_t *lock)
+{
+    /* It is possible that on x86 the atomic in the lock is strong enough
+	 * to force-flush the WC buffers quickly, and this SFENCE can be
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: alignment in comment

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 8]
```diff
@@ -6,6 +6,312 @@
...
+    return XCCL_OK;
+}
+
+static inline void post_send_db(struct internal_qp *mqp, int nreq, void *ctrl)
+{
+    if (ucs_unlikely(!nreq))
+        return;
+
+    /*
+	 * Make sure that descriptors are written before
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: alignment

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 9]
```diff
@@ -6,6 +6,312 @@
...
+common_wqe_init(struct ibv_qp_ex *ibqp, struct internal_qp *mqp)
+{
+    struct mlx5_wqe_ctrl_seg *ctrl;
+    uint8_t fence;
+    uint32_t idx;
+
+    idx = mqp->sq_cur_post & (mqp->qp.sq.wqe_cnt - 1);
+
+    ctrl = mqp->sq_start + (idx << MLX5_SEND_WQE_SHIFT);
+    *(uint32_t *)((void *)ctrl + 8) = 0;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: void * arithmetics

File: src/team_lib/mhba/xccl_mhba_ib.c [hunk 10]
```diff
@@ -6,6 +6,312 @@
...
+                                      IBV_ACCESS_LOCAL_WRITE |
+                                              IBV_ACCESS_REMOTE_WRITE |
+                                              IBV_ACCESS_REMOTE_READ |
+                                              IBV_ACCESS_REMOTE_ATOMIC))) {
+        xccl_mhba_error("Un-supported UMR flags");
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    common_wqe_init(ibqp, mqp);
+    mqp->cur_size      = sizeof(struct mlx5_wqe_ctrl_seg) / 16;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.c]: what is 16 here? if some constant plz give it a name for readability

File: src/team_lib/mhba/xccl_mhba_ib.h [hunk 11]
```diff
@@ -6,7 +6,52 @@
...
+MAKE_WRITE(mmio_write32, 32)
+
+/* UMR pointer to KLMs/MTTs/RepeatBlock and BSFs location (when inline = 0) */
+struct mlx5_wqe_umr_pointer_seg {
+    __be32 reserved;
+    __be32 mkey;
+    __be64 address;
+};
+
+struct internal_qp;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.h]: name prefix? xccl_mhba?

File: src/team_lib/mhba/xccl_mhba_lib.h [hunk 12]
```diff
@@ -107,20 +107,47 @@ typedef struct xccl_mhba_op {
...
-    struct mlx5dv_qp_ex *umr_mlx5dv_qp_ex;
+    int                      asr_rank;
+    xccl_sbgp_t             *sbgp;
+    void                    *storage;
+    xccl_mhba_op_t           ops[MAX_OUTSTANDING_OPS];
+    struct mlx5dv_mkey      *team_recv_mkey;
+    struct ibv_context      *shared_ctx;
+    struct ibv_pd           *shared_pd;
+    struct ibv_cq           *umr_cq;
+    struct xccl_mhba_mlx5_qp ns_umr_qp;
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_lib.h]: plz add some commentary here: e.g. "ns_umr_qp" - used for ..., s_umr_qp - used_for ...

File: src/team_lib/mhba/xccl_mhba_mkeys.c [hunk 13]
```diff
@@ -157,29 +182,22 @@ static xccl_status_t poll_umr_cq(xccl_mhba_node_t *node)
 }
 
 // Execute the UMR WQE for populating the UMR's MasterMKey
-static xccl_status_t populate_mkey(xccl_mhba_team_t *team, int mem_access_flags,
-                                   struct mlx5dv_mkey *mkey, void *mkey_entries,
-                                   int repeat_count, int strided)
+static xccl_status_t
+populate_non_strided_mkey(xccl_mhba_team_t *team, int mem_access_flags,
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_mkeys.c:186]: For non_strided do we still have the INLINE UMR limit?

File: src/team_lib/mhba/xccl_mhba_team.c [hunk 14]
```diff
@@ -495,10 +492,31 @@ xccl_status_t xccl_mhba_team_create_post(xccl_tl_context_t  *context,
...
         memset(mhba_team->cq_completions, 0, sizeof(mhba_team->cq_completions));
+        /* allocate buffer for noninline UMR registration, has to be 2KB aligned */
+        umr_buf_size = align(sizeof(struct mlx5_wqe_umr_repeat_ent_seg) *
+                                     (mhba_team->node.sbgp->group_size + 1),
+                             64);
+        if (posix_memalign(&mhba_team->node.umr_entries_buf, 2048,
+                           umr_buf_size) != 0) {
+            xccl_mhba_error(
+                    "failed to allocate %zu bytes for noninline UMR buffer: %m",
+                    sizeof(struct mlx5_wqe_umr_pointer_seg) *
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_team.c]: you have 2 format specifiers and only 1 value provided. additionally you can use umr_buf_size as argument

File: src/team_lib/mhba/xccl_mhba_ib.h [hunk 15]
```diff
@@ -6,7 +6,52 @@
...
+
+#define MAKE_WRITE(_NAME_, _SZ_)                                               \
+    static inline void _NAME_##_be(void *addr, __be##_SZ_ value)               \
+    {                                                                          \
+        atomic_store_explicit((_Atomic(uint##_SZ_##_t) *)addr,                 \
+                              (__force uint##_SZ_##_t)value,                   \
+                              memory_order_relaxed);                           \
+    }
+
+MAKE_WRITE(mmio_write32, 32)
```
Comment (vspetrov) [inline src/team_lib/mhba/xccl_mhba_ib.h:35]: Unused?

Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_ib.h:35]: Used - it calls the define on top of it to define the needed function

File: src/team_lib/mhba/xccl_mhba_mkeys.c (hunk 13 above)
Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_mkeys.c:186]: Yes - where is this relevant in the code?

File: src/team_lib/mhba/xccl_mhba_ib.c (hunk 5 above)
Comment (lappazos) [reply src/team_lib/mhba/xccl_mhba_ib.c:64]: kept as TODO for the future

**Review by vspetrov — APPROVED** [review]

**Pull request merged**